torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py

```diff
@@ -243,6 +246,68 @@ def aot_dispatch_base(
     return compiled_fn
 
 
+def collect_fw_donated_buffer_idxs(
+    fw_ins: List[Optional[FakeTensor]],
+    user_fw_outs: List[Optional[FakeTensor]],
+    bw_outs: List[Optional[FakeTensor]],
+    saved_tensors: List[FakeTensor],
+) -> List[int]:
+    """
+    Checks if the saved tensors are donated buffers, which means a saved tensor is not
+    an alias of any tensors in fw_ins, user_fw_outs, and bw_outs.
+    """
+
+    storage_refs = set()
+    for t in itertools.chain(fw_ins, user_fw_outs, bw_outs):
+        if isinstance(t, FakeTensor):
+            storage_refs.add(StorageWeakRef(t.untyped_storage()))
+
+    num_saved_tensor = len(saved_tensors)
+    donated_buffer_idxs = []
+    for i in range(num_saved_tensor):
+        t = saved_tensors[i]
+        if StorageWeakRef(t.untyped_storage()) not in storage_refs:
+            donated_buffer_idxs.append(i)
+
+    return donated_buffer_idxs
+
+
+def collect_bw_donated_buffer_idxs(
+    fw_module: torch.fx.GraphModule,
+    bw_module: torch.fx.GraphModule,
+    fw_metadata: ViewAndMutationMeta,
+) -> List[int]:
+    """
+    Collects backward donated buffer indexes from fw_module and bw_module.
+    """
+
+    fw_ins = fw_module.graph.find_nodes(op="placeholder")
+    bw_outs = next(reversed(bw_module.graph.find_nodes(op="output"))).args[0]
+    fw_outs = next(reversed(fw_module.graph.find_nodes(op="output"))).args[0]
+
+    # check if every node has meta["val"] since meta["val"] may be lost during
+    # graph transformation.
+    try:
+        fw_ins = [n.meta["val"] if hasattr(n, "meta") else None for n in fw_ins]
+        fw_outs = [n.meta["val"] if hasattr(n, "meta") else None for n in fw_outs]
+        bw_outs = [n.meta["val"] if hasattr(n, "meta") else None for n in bw_outs]
+    except KeyError:
```

Comment:
why are we wrapping this in a try/except? When can it fail?

Reply:
hmm I don't think so (at least, inside of AOTAutograd). I think if that did somehow happen I would rather we raise a loud error, since that would probably be a bug